### DISCUSSION TITLE
supported language

### DISCUSSION THREAD

**Discussion by setop** [firebuild/firebuild#1196] [Ideas]
What are the supported languages apart from c/c+= and rust ? What would be the requirements to support new languages ?

I'm especially interested in Elixir (with Mix build tool).

**Comment by rbalint** [comment]
There are many supported languages/compilers: https://firebuild.com/#about .
Firebuild in general supports caching and quickly replaying "well behaving" processes that generate input based on the environment and the input files.

Typical C and similar compilers work like that and `make` and other build systems drive those steps from a separate process. 
Erlang's `emake` and `mix` can't be accelerated at the moment because the Erlang family of languages tend to run both the build system and the compilation steps in the same `beam.smp` system process, thus firebuild can't detect the information boundaries for the compilation steps. JavaScript build systems have the same problem.
You can observe this behavior by running `firebuild -r mix compile` in your project, see that `beam.smp` does most of the work and there are not many other system processes.

The solution would be extending `mix` to not rely on firebuild's per process standard input and output detection, but to report and request shortcutting smaller well defined work items inside the bigger `beam.smp` process. That scheme could be used for JavaScript compilation steps, too, as a quick, generic language-independent caching framework.
